File: extensions/positron-python/uitests/uitests/steps/problems.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.


import behave
from selenium.common.exceptions import StaleElementReferenceException

import uitests.vscode.status_bar
import uitests.tools
import uitests.vscode.problems


@behave.then("there are no problems in the problems panel")
@uitests.tools.retry((AssertionError, StaleElementReferenceException))
def then_no_problems(context):
    count = uitests.vscode.problems.get_problem_count(context)
    assert count == 0, f"Number of problems is {count}"


@behave.then("there is at least one problem in the problems panel")
@uitests.tools.retry((AssertionError, StaleElementReferenceException))
def then_atleast_one_problem(context):
    count = uitests.vscode.problems.get_problem_count(context)
    assert count > 0


@behave.then("there are at least {problem_count:Number} problems in the problems panel")
@uitests.tools.retry((AssertionError, StaleElementReferenceException))
def then_atleast_n_problems(context, problem_count):
    count = uitests.vscode.problems.get_problem_count(context)
    assert problem_count >= count, f"{problem_count} should be >= {count}"


@behave.then('there is a problem with the message "{message}"')
@uitests.tools.retry((AssertionError, StaleElementReferenceException))
def then_has_problem_with_message(context, message):
    messages = uitests.vscode.problems.get_problems(context)
    all_problems = "".join(messages)
    assert message in all_problems, f"{message} not in {all_problems}"


@behave.then('there is a problem with the file named "{name}"')
@uitests.tools.retry((AssertionError, StaleElementReferenceException))
def then_has_poroblem_in_filename(context, name):
    files = uitests.vscode.problems.get_problem_files(context)
    all_files = "".join(files)
    assert name in all_files, f"{all_files} not in {all_files}"

File: extensions/positron-python/uitests/src/steps/problems.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

'use strict';

// tslint:disable: no-invalid-this

import * as assert from 'assert';
import { expect } from 'chai';
import { Then } from 'cucumber';
import { CucumberRetryMax5Seconds } from '../constants';

// Wait for some time as it take take at least 1s to appear.
// Surely problems won't take more than 5 seconds to appear.
// Why 5? Well, needs to be > 1, but most certainly not more than 5.

Then('there are no problems in the problems panel', CucumberRetryMax5Seconds, async function() {
    const count = await this.app.problems.getProblemCount();
    assert.equal(count, 0);
});

Then('there is at least one problem in the problems panel', CucumberRetryMax5Seconds, async function() {
    const count = await this.app.problems.getProblemCount();
    expect(count).to.greaterThan(0);
});

Then('there are at least {int} problems in the problems panel', CucumberRetryMax5Seconds, async function(expectedMinimumCount: number) {
    const count = await this.app.problems.getProblemCount();
    expect(count).to.greaterThan(expectedMinimumCount - 1);
});

Then('there is a problem with the message {string}', CucumberRetryMax5Seconds, async function(message: string) {
    const messages = await this.app.problems.getProblemMessages();
    expect(messages.join(', ').toLowerCase()).to.include(message.toLowerCase());
});

Then('there is a problem with the file named {string}', CucumberRetryMax5Seconds, async function(fileName: string) {
    const messages = await this.app.problems.getProblemFiles();
    expect(messages.join(', ').toLowerCase()).to.include(fileName.toLowerCase());
});
